Can't find repository_id in []

The following query:

```sql
SELECT
                    COUNT(*)
                FROM (
                SELECT DISTINCT
                     s.repository_id AS repository_id
                FROM (
                    SELECT
                        commit_hash,
                        MAX(STRUCT(history_index, repository_id)) AS s
                    FROM ref_commits
                    NATURAL JOIN commits
                    WHERE
                        history_index != 1
                        AND ref_name LIKE 'refs/heads/HEAD/%'
                        AND SIZE(PARSE_COMMIT_PARENTS(commit_parents)) == 0
                    GROUP BY commit_hash
                ) AS q
                ) AS q2
```

produces the following error:

```
Exchange SinglePartition
+- *(3) HashAggregate(keys=[], functions=[partial_count(1)], output=[count#104L])
   +- *(3) HashAggregate(keys=[repository_id#8], functions=[], output=[])
      +- Exchange hashpartitioning(repository_id#8, 200)
         +- *(2) HashAggregate(keys=[repository_id#8], functions=[], output=[repository_id#8])
            +- *(2) HashAggregate(keys=[commit_hash#9], functions=[], output=[])
               +- Exchange hashpartitioning(commit_hash#9, 200)
                  +- *(1) HashAggregate(keys=[commit_hash#9], functions=[], output=[commit_hash#9])
                     +- *(1) Project [commit_hash#9]
                        +- *(1) Filter (size(UDF:parse_commit_parents(commit_parents#26)) = 0)
                           +- *(1) DataSourceV2Scan [commit_hash#9, commit_parents#26], DefaultReader(List(GitbaseServer(localhost:3308,root,)),StructType(StructField(repository_id,StringType,true), StructField(commit_hash,StringType,true), StructField(commit_parents,BinaryType,true)),Project(List(repository_id#8, commit_hash#9, commit_parents#26),Join(Table(ref_commits),Table(commits),Some(((repository_id#8 = repository_id#16) && (commit_hash#9 = commit_hash#17))))))

org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:
Exchange SinglePartition
+- *(3) HashAggregate(keys=[], functions=[partial_count(1)], output=[count#104L])
   +- *(3) HashAggregate(keys=[repository_id#8], functions=[], output=[])
      +- Exchange hashpartitioning(repository_id#8, 200)
         +- *(2) HashAggregate(keys=[repository_id#8], functions=[], output=[repository_id#8])
            +- *(2) HashAggregate(keys=[commit_hash#9], functions=[], output=[])
               +- Exchange hashpartitioning(commit_hash#9, 200)
                  +- *(1) HashAggregate(keys=[commit_hash#9], functions=[], output=[commit_hash#9])
                     +- *(1) Project [commit_hash#9]
                        +- *(1) Filter (size(UDF:parse_commit_parents(commit_parents#26)) = 0)
                           +- *(1) DataSourceV2Scan [commit_hash#9, commit_parents#26], DefaultReader(List(GitbaseServer(localhost:3308,root,)),StructType(StructField(repository_id,StringType,true), StructField(commit_hash,StringType,true), StructField(commit_parents,BinaryType,true)),Project(List(repository_id#8, commit_hash#9, commit_parents#26),Join(Table(ref_commits),Table(commits),Some(((repository_id#8 = repository_id#16) && (commit_hash#9 = commit_hash#17))))))

	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.doExecute(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.InputAdapter.inputRDDs(WholeStageCodegenExec.scala:371)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.inputRDDs(HashAggregateExec.scala:150)
	at org.apache.spark.sql.execution.BaseLimitExec$class.inputRDDs(limit.scala:62)
	at org.apache.spark.sql.execution.LocalLimitExec.inputRDDs(limit.scala:97)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:605)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan.getByteArrayRdd(SparkPlan.scala:247)
	at org.apache.spark.sql.execution.SparkPlan.executeTake(SparkPlan.scala:337)
	at org.apache.spark.sql.execution.CollectLimitExec.executeCollect(limit.scala:38)
	at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$collectFromPlan(Dataset.scala:3273)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2484)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2484)
	at org.apache.spark.sql.Dataset$$anonfun$52.apply(Dataset.scala:3254)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:77)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3253)
	at org.apache.spark.sql.Dataset.head(Dataset.scala:2484)
	at org.apache.spark.sql.Dataset.take(Dataset.scala:2698)
	at org.apache.spark.sql.Dataset.showString(Dataset.scala:254)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:725)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:702)
	at tech.sourced.gitbase.spark.DefaultSourceSpec$$anonfun$13.apply$mcV$sp(DefaultSourceSpec.scala:377)
	at tech.sourced.gitbase.spark.DefaultSourceSpec$$anonfun$13.apply(DefaultSourceSpec.scala:377)
	at tech.sourced.gitbase.spark.DefaultSourceSpec$$anonfun$13.apply(DefaultSourceSpec.scala:377)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FlatSpecLike$$anon$1.apply(FlatSpecLike.scala:1682)
	at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
	at org.scalatest.FlatSpec.withFixture(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$class.invokeWithFixture$1(FlatSpecLike.scala:1679)
	at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
	at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
	at org.scalatest.FlatSpecLike$class.runTest(FlatSpecLike.scala:1692)
	at org.scalatest.FlatSpec.runTest(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
	at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:373)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:410)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
	at org.scalatest.FlatSpecLike$class.runTests(FlatSpecLike.scala:1750)
	at org.scalatest.FlatSpec.runTests(FlatSpec.scala:1685)
	at org.scalatest.Suite$class.run(Suite.scala:1147)
	at org.scalatest.FlatSpec.org$scalatest$FlatSpecLike$$super$run(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
	at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
	at org.scalatest.FlatSpecLike$class.run(FlatSpecLike.scala:1795)
	at tech.sourced.gitbase.spark.DefaultSourceSpec.org$scalatest$BeforeAndAfterAll$$super$run(DefaultSourceSpec.scala:4)
	at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
	at tech.sourced.gitbase.spark.DefaultSourceSpec.run(DefaultSourceSpec.scala:4)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1346)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1340)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1340)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1011)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1010)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1506)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1010)
	at org.scalatest.tools.Runner$.run(Runner.scala:850)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:131)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)
Caused by: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:
Exchange hashpartitioning(repository_id#8, 200)
+- *(2) HashAggregate(keys=[repository_id#8], functions=[], output=[repository_id#8])
   +- *(2) HashAggregate(keys=[commit_hash#9], functions=[], output=[])
      +- Exchange hashpartitioning(commit_hash#9, 200)
         +- *(1) HashAggregate(keys=[commit_hash#9], functions=[], output=[commit_hash#9])
            +- *(1) Project [commit_hash#9]
               +- *(1) Filter (size(UDF:parse_commit_parents(commit_parents#26)) = 0)
                  +- *(1) DataSourceV2Scan [commit_hash#9, commit_parents#26], DefaultReader(List(GitbaseServer(localhost:3308,root,)),StructType(StructField(repository_id,StringType,true), StructField(commit_hash,StringType,true), StructField(commit_parents,BinaryType,true)),Project(List(repository_id#8, commit_hash#9, commit_parents#26),Join(Table(ref_commits),Table(commits),Some(((repository_id#8 = repository_id#16) && (commit_hash#9 = commit_hash#17))))))

	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.doExecute(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.InputAdapter.inputRDDs(WholeStageCodegenExec.scala:371)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.inputRDDs(HashAggregateExec.scala:150)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.inputRDDs(HashAggregateExec.scala:150)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:605)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.prepareShuffleDependency(ShuffleExchangeExec.scala:92)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:128)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	... 86 more
Caused by: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: Binding attribute, tree: repository_id#8
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1.applyOrElse(BoundAttribute.scala:91)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1.applyOrElse(BoundAttribute.scala:90)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$2.apply(TreeNode.scala:267)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$2.apply(TreeNode.scala:267)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:70)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:266)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transform(TreeNode.scala:256)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$.bindReference(BoundAttribute.scala:90)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec$$anonfun$48.apply(HashAggregateExec.scala:737)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec$$anonfun$48.apply(HashAggregateExec.scala:737)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:296)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.doConsumeWithKeys(HashAggregateExec.scala:737)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.doConsume(HashAggregateExec.scala:173)
	at org.apache.spark.sql.execution.CodegenSupport$class.constructDoConsumeFunction(WholeStageCodegenExec.scala:208)
	at org.apache.spark.sql.execution.CodegenSupport$class.consume(WholeStageCodegenExec.scala:179)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.consume(HashAggregateExec.scala:39)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.generateResultFunction(HashAggregateExec.scala:515)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.doProduceWithKeys(HashAggregateExec.scala:649)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.doProduce(HashAggregateExec.scala:165)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:88)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.CodegenSupport$class.produce(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.produce(HashAggregateExec.scala:39)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.doProduceWithKeys(HashAggregateExec.scala:641)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.doProduce(HashAggregateExec.scala:165)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:88)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.CodegenSupport$class.produce(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.aggregate.HashAggregateExec.produce(HashAggregateExec.scala:39)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doCodeGen(WholeStageCodegenExec.scala:524)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:576)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.prepareShuffleDependency(ShuffleExchangeExec.scala:92)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:128)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	... 107 more
Caused by: java.lang.RuntimeException: Couldn't find repository_id#8 in []
	at scala.sys.package$.error(package.scala:27)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1$$anonfun$applyOrElse$1.apply(BoundAttribute.scala:97)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1$$anonfun$applyOrElse$1.apply(BoundAttribute.scala:91)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	... 158 more
```

--- a/src/main/scala/tech/sourced/gitbase/spark/rule/PushdownJoins.scala
+++ b/src/main/scala/tech/sourced/gitbase/spark/rule/PushdownJoins.scala
@@ -9,7 +9,7 @@ import org.apache.spark.sql.catalyst.plans.logical
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 import org.apache.spark.sql.catalyst.rules.Rule
 import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
-import org.apache.spark.sql.types.{Metadata, StructField, StructType}
+import org.apache.spark.sql.types.Metadata
 import tech.sourced.gitbase.spark._
 
 object PushdownJoins extends Rule[LogicalPlan] {

--- a/src/main/scala/tech/sourced/gitbase/spark/rule/package.scala
+++ b/src/main/scala/tech/sourced/gitbase/spark/rule/package.scala
@@ -1,10 +1,8 @@
 package tech.sourced.gitbase.spark
 
 import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeReference}
-import org.apache.spark.sql.catalyst.plans.logical
 import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
 import org.apache.spark.sql.catalyst.rules.Rule
-import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
 import org.apache.spark.sql.types.{StructField, StructType}
 
 package object rule {
@@ -40,18 +38,7 @@ package object rule {
 
   private[rule] def fixAttributeReferences(plan: LogicalPlan): LogicalPlan = {
     import JoinOptimizer._
-    val availableAttrs: Seq[Attribute] = plan.children.flatMap(child => {
-      child.find {
-        case _: logical.Project => true
-        case DataSourceV2Relation(_, _: DefaultReader) => true
-        case _: logical.Join => true
-        case _ => false
-      } match {
-        case Some(logical.Project(attrs, _)) => attrs.map(_.toAttribute)
-        case Some(DataSourceV2Relation(output, _: DefaultReader)) => output
-        case _ => Seq()
-      }
-    })
+    val availableAttrs: Seq[Attribute] = plan.children.flatMap(_.output)
 
     plan.transformExpressionsUp {
       case a: Attribute =>

--- a/src/test/scala/tech/sourced/gitbase/spark/DefaultSourceSpec.scala
+++ b/src/test/scala/tech/sourced/gitbase/spark/DefaultSourceSpec.scala
@@ -247,7 +247,7 @@ class DefaultSourceSpec extends BaseGitbaseSpec {
         |WHERE
         |  num > 1
       """.stripMargin)
-    df.show(20, false)
+    df.show(20, truncate = false)
   }
 
   it should "approximately detect forks" in {
@@ -277,6 +277,64 @@ class DefaultSourceSpec extends BaseGitbaseSpec {
     result should equal(Array(
       ("fff7062de8474d10a67d417ccea87ba6f58ca81d", 2),
       ("fff840f8784ef162dc83a1465fc5763d890b68ba", 2)
+    ))
+  }
+
+  it should "count HEADs excluding forks" in {
+    val result = spark.sql(
+      """SELECT
+                    COUNT(*)
+                FROM (
+                  SELECT DISTINCT
+                       s.repository_id AS repository_id
+                  FROM (
+                      SELECT
+                          commit_hash,
+                          MAX(STRUCT(history_index, repository_id)) AS s
+                      FROM ref_commits
+                      NATURAL JOIN commits
+                      WHERE
+                          history_index != 1
+                          AND ref_name LIKE 'refs/heads/HEAD/%'
+                          AND SIZE(PARSE_COMMIT_PARENTS(commit_parents)) == 0
+                      GROUP BY commit_hash
+                  ) AS q
+                ) AS q2""").collect()(0)(0)
+    result should be(3L)
+  }
+
+  it should "do repository count by language presence" in {
+    val result = spark.sql(
+      """SELECT
+                    language,
+                    COUNT(repository_id) AS repository_count
+                FROM (
+                    SELECT DISTINCT
+                        repository_id,
+                        COALESCE(language(file_path, blob_content), 'Unknown') AS language
+                    FROM ref_commits
+                    NATURAL JOIN commit_files
+                    NATURAL JOIN files
+                    WHERE
+                        ref_name LIKE 'refs/heads/HEAD/%'
+                    ) AS q2
+                GROUP BY language
+                ORDER BY repository_count DESC, language ASC""").collect().map(r => (r(0), r(1)))
+
+    result should equal(Array(
+      ("Text", 3),
+      ("", 2),
+      ("Markdown", 2),
+      ("C", 1),
+      ("C++", 1),
+      ("CMake", 1),
+      ("JSON", 1),
+      ("JavaScript", 1),
+      ("QML", 1),
+      ("Ruby", 1),
+      ("Shell", 1),
+      ("XML", 1),
+      ("desktop", 1)
     ))
   }
 

